hal-x86_64/src/mm/mod.rs

```diff
@@ -66,26 +68,33 @@ pub fn init_paging(vm_offset: VAddr, recursive_addr: VAddr) {
         }
     }
     tracing::trace!(present_entries);
-    let pml4_paddr = PAddr::from_usize((vm_offset - recursive_addr).as_usize());
-    tracing::debug!(?pml4_paddr);
 
-    let pml4_page =
-        Page::<PAddr, Size4Kb>::starting_at(pml4_paddr).expect("PML4 not aligned, what the hell!");
-    let pml4_2 = unsafe { &mut *recursive_addr.as_ptr::<PageTable<level::Pml4>>() };
-    tracing::debug!("dereferenced pml4");
-    let mut present_entries = 0;
-    for (idx, entry) in (&pml4_2.entries[..]).iter().enumerate() {
-        // if entry.is_present() {
-        tracing::trace!(idx, ?entry);
-        //     present_entries += 1;
-        // }
-    }
-    tracing::info!(present_entries);
+    // Create the recursive pagetable entry.
+    assert!(
+        !pml4.entries[RECURSIVE_INDEX].is_present(),
+        "bootloader must not have used entry 511"
```

Comment:
should we just clobber it if it did?

Reply:
I don't think so, as that page entry inserted by the bootloader could be for the stack or program code, which would be very unfortunate to replace with a recursive pagetable entry.
I think it's better to crash than to risk that.

Reply:
yeah, we'd have to remap that entry somewhere else first.  eventually i _would_ rather not be panicking at boot-time due to (possibly non-deterministic?) bootloader behavior, but it doesn't _seem_ to ever want to use 511 in qemu tests, so, for now, who cares!